04-Protocol-Security.md

```diff
@@ -0,0 +1,391 @@
+Prior to opening any Stratum V2 channels for mining, clients MUST first initiate the cryptographic session state that is used to encrypt all messages sent between themselves and servers.
+Thus, the cryptographic session state is independent of V2 messaging conventions.
+
+At the same time, this specification proposes optional use of a particular handshake protocol based on the **[Noise Protocol framework](https://noiseprotocol.org/noise.html)**.
+The client and server establish secure communication using Diffie-Hellman (DH) key agreement, as described in greater detail in the Authenticated Key Agreement Handshake section below.
+
+Using the handshake protocol to establish secured communication is **optional** on the local network (e.g. local mining devices talking to a local mining proxy).
+However, it is **mandatory** for remote access to the upstream nodes, whether they be pool mining services, job negotiating services or template distributors.
+
+
+## 4.1 Motivation for Authenticated Encryption with Associated Data
+Data transferred by the mining protocol MUST not provide adversary information that they can use to estimate the performance of any particular miner. Any intelligence about submitted shares can be directly converted to estimations of a miner’s earnings and can be associated with a particular username. This is unacceptable privacy leakage that needs to be addressed.
+
+
+## 4.2 Motivation for Using the Noise Protocol Framework
+The reasons why Noise Protocol Framework has been chosen are listed below:
+
+- The Framework provides a formalism to describe the handshake protocol that can be verified.
+- A custom certificate scheme is now possible (no need to use x509 certificates).
+
+## 4.3 Choice of cryptographic primitives
+
+Noise encrypted session requires Elliptic Curve (EC), Hash function (`HASH()`) and cipher function that supports AEAD mode<sup>[1](#reference-1)</sup>.
+
+This specification describes mandatory cryptographic primitives that each implementation needs to support.
+These primitives are chosen so that Noise Encryption layer for Stratum V2 can be implemented using primitives already present in Bitcoin Core project at the time of writing this spec.
+
+### 4.3.1 Elliptic Curve
+* Bitcoin's secp256k1 curve<sup>[2](#reference-2)</sup> is used
+* Schnorr signature scheme is used as described in BIP340<sup>[3](#reference-3)</sup>
+
+#### 4.3.1.1 EC point encoding remarks
+Secp256k1 curve points, which includes Public Keys and ECDH results, are points with of X- and Y-coordinate, 32-bytes each. There are several possibilities how to serialize them:
+1. 64-byte full X- and Y-coordinate serialization for public keys (and ECDH results) and 96 Bytes for signatures.
+2. 33-byte X-coordinate with 1 parity bit serialization for public keys and similarly 65-byte for signatures.
+3. 32-byte X-coordinate only with implicit Y-coordinate for public keys and 64-byte for signatures.
```

Comment:
Maybe list 3 as 1 since this is the one we are choosing to implement.

Reply:
I'd keep it on the 3rd position. Because it's the *least obvious* approach.

Reply:
You might want to consider using option 1 given the secp256k1 C-lib used in core expects 64 bytes for ecdh operations. Not sure if it accepts 32 byte x-coordinate plus padding, would need to be investigated. Also the Rust secp256k1 library is an ffi wrapper that calls into the same C-lib secp256k1 and it seems like the PublicKey wrapper is also 64 bytes when calling ecdh operation:

- [Rust secp256k1 - SharedSecret and call to secp256k1_ecdh](https://github.com/rust-bitcoin/rust-secp256k1/blob/master/src/ecdh.rs#L48)
- [Rust secp256k1 - PublicKey (64 bytes)](https://github.com/rust-bitcoin/rust-secp256k1/blob/master/secp256k1-sys/src/lib.rs#L136)
- [C secp256k1 - secp256k1_ecdh()](https://github.com/bitcoin/bitcoin/blob/master/src/secp256k1/include/secp256k1_ecdh.h#L36)
- [C secp256k1 - secp256k1_pubkey](https://github.com/bitcoin/bitcoin/blob/master/src/secp256k1/include/secp256k1.h#L61)

Reply:
This is about representation *on the wire*
The full public key as EC point is always reconstructed from the X coordinate assuming it has even parity.